pkg/cloudprovider/aws/object_storage_adapter.go

```diff
@@ -27,14 +27,16 @@ import (
 var _ cloudprovider.ObjectStorageAdapter = &objectStorageAdapter{}
 
 type objectStorageAdapter struct {
-	s3 *s3.S3
+	s3       *s3.S3
+	KMSKeyID string
 }
 
 func (op *objectStorageAdapter) PutObject(bucket string, key string, body io.ReadSeeker) error {
 	req := &s3.PutObjectInput{
-		Bucket: &bucket,
-		Key:    &key,
-		Body:   body,
+		Bucket:      &bucket,
+		Key:         &key,
+		SSEKMSKeyId: &op.KMSKeyID,
```

Comment:
I find the docs unclear regarding `GetObjectInput` with KMS. Deleting and listing objects is definitely possible without the key. I compared the code with the Docker registries implementation and usage of GetObject and it seems that `s3.GetObject` does "the right thing" if it tries to get an encrypted object.